rfcs/012 - Either, Option, and Result Destructuring.md

```diff
@@ -0,0 +1,110 @@
+# 012 - Easier Either, Option, and Result Destructuring
+
+## Current Status
+
+### Proposed
+
+2021-01-08
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- Full Name <email@example.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/alantech/alan/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- Colton Donnelly <colton@alantechnologies.com>
+
+## Summary
+
+Currently, error handling in Alan is done through the `Result<T>` type, which is based on Rust's `Result<T, E>` type.
+While Rust has convenient ways of retrieving the value of a `Result`, it's difficult to replicate this behavior due to a lack of pattern matching.
+This RFC seeks to introduce a new language feature that would allow Alan to replicate the convenience of Rust's `Result` destructuring without introducing pattern matching.
+
+## Expected SemBDD Impact
+
+If Alan was at least version 1.0.0, then this would be considered a major change that breaks user code when destructuring a `Result<T>`.
+
+## Proposal
+
+This proposal seeks to introduce a special case when handling `if` statements in Alan.
+Currently, when handling a `Result<T>`, a user might expect to write something like this in order to retrieve a value.
+```
+let myResult = ok(5);
+if myResult.isOk() {
+	let okVal = myResult.getOr(-1);
+	print("should be 5: " + okVal);
+} else {
+	let errVal = myResult.getErr("expected error, but could not retrieve Error value");
+	print("unreachable, but I know that errVal is NOT 'expected error, but could not retrieve Error value', unless that's what myResult is");
+}
+```
+
+Note that, in both branches of the above `if` statement, the status (`ok` or `err`) of the `Result` is already known, but the user must then call either `getOr` or `getErr` in order to retrieve the expected value.
+(Note that `getOrDefault` is not yet implemented, but only makes this process marginally easier.)
```

Comment:
`getOrExit` could also be used and exists now, but it does imply that the user intentionally wants to crash their app if there was an error, which is semantically incorrect in this case.